Added documentation for System.Text.Rune struct

jozkee opened the pull request:
Documented APIs:
* op_Equality(System.Text.Rune,System.Text.Rune)
* op_Explicit(System.UInt32)~System.Text.Rune
* op_Explicit(System.Char)~System.Text.Rune
* op_Explicit(System.Int32)~System.Text.Rune
* op_GreaterThan(System.Text.Rune,System.Text.Rune)
* op_GreaterThanOrEqual(System.Text.Rune,System.Text.Rune)
* op_Inequality(System.Text.Rune,System.Text.Rune)
* op_LessThan(System.Text.Rune,System.Text.Rune)
* op_LessThanOrEqual(System.Text.Rune,System.Text.Rune)
* ToLower(System.Text.Rune,System.Globalization.CultureInfo)
* ToLowerInvariant(System.Text.Rune)
* ToUpper(System.Text.Rune,System.Globalization.CultureInfo)
* ToUpperInvariant(System.Text.Rune)

Jozkee commented:
@carlossanlop @rpetrusha @mairaw

Jozkee commented:
@GrabYourPitchforks note that I am referring to `Rune` as a 32-bit Unicode character. Please correct me if I am wrong or if we should omit that.

carlossanlop commented on xml/System.Text/Rune.xml:
```diff
@@ -977,9 +977,9 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="ch">The 16-bit Unicode character to convert.</param>
+        <summary>Defines an explicit conversion of a 16-bit Unicode character to a <see cref="T:System.Text.Rune"/>.</summary>
```
The summary could be improved. See this op_Explicit description: https://docs.microsoft.com/en-us/dotnet/api/system.intptr.op_explicit?view=netcore-3.0

carlossanlop replied:
Same suggestion for the other op_Explicit below.

Jozkee replied:
You mean using similar words to IntPtr.Explicit summary?

I used [Decimal.Explicit](https://docs.microsoft.com/en-us/dotnet/api/system.decimal.op_explicit?view=netcore-3.0) as a base, but I can switch it if you think is better your proposal.

rpetrusha replied:
I think that the language from Decimal.Explicit ("Defines an explicit conversion...") is better, given that an explicit conversion requires a casting operator -- it performs a conversion only in a qualified or conditional sense.

carlossanlop replied:
No problem then, let's keep the current language.

carlossanlop commented on xml/System.Text/Rune.xml:
```diff
@@ -1058,10 +1058,10 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="left">The first value to compare.</param>
+        <param name="right">the second value to compare.</param>
```
Capital T:
```suggestion
        <param name="right">The second value to compare.</param>
```

carlossanlop commented on xml/System.Text/Rune.xml:
```diff
@@ -1085,10 +1085,10 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="left">The first value to compare.</param>
+        <param name="right">the second value to compare.</param>
```
Capital T:
```suggestion
        <param name="right">The second value to compare.</param>
```

carlossanlop commented on xml/System.Text/Rune.xml:
```diff
@@ -1112,10 +1112,10 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="left">The first value to compare.</param>
+        <param name="right">the second value to compare.</param>
```
Capital T:
```suggestion
        <param name="right">The second value to compare.</param>
```

carlossanlop commented on xml/System.Text/Rune.xml:
```diff
@@ -1166,10 +1166,10 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="left">The first value to compare.</param>
+        <param name="right">the second value to compare.</param>
```
```suggestion
        <param name="right">The second value to compare.</param>
```

carlossanlop commented on xml/System.Text/Rune.xml:
```diff
@@ -1166,10 +1166,10 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
+        <param name="left">The first value to compare.</param>
+        <param name="right">the second value to compare.</param>
+        <summary>Returns a value indicating whether a specified <see cref="T:System.Text.Rune"/> is less than or equal to another specified <see cref="T:System.Text.Rune"/>.</summary>
+        <returns><see langword="true" /> if <paramref name="left" /> is less than or equals to <paramref name="right" />; otherwise, <see langword="false" />.</returns>
```
```suggestion
        <returns><see langword="true" /> if <paramref name="left" /> is less than or equal to <paramref name="right" />; otherwise, <see langword="false" />.</returns>
```

rpetrusha commented on xml/System.Text/Rune.xml:
```diff
@@ -977,9 +977,9 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
+        <summary>Defines an explicit conversion of a 16-bit Unicode character to a <see cref="T:System.Text.Rune"/>.</summary>
+        <returns><paramref name="ch" /> as a 32-bit Unicode character.</returns>
         <remarks>To be added.</remarks>
       </Docs>
```
This should document the ArgumentOutOfRangeException for a surrogate.

rpetrusha commented on xml/System.Text/Rune.xml:
```diff
@@ -1002,9 +1002,9 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
+        <param name="value">The 32-bit signed integer to convert.</param>
+        <summary>Defines an explicit conversion of a 32-bit signed integer to a <see cref="T:System.Text.Rune"/>.</summary>
+        <returns><paramref name="value" /> as a 32-bit Unicode character.</returns>
         <remarks>To be added.</remarks>
```
Should document the ArgumentOutOfRangeException for an invalid Unicode scalar value.

Jozkee replied:
```suggestion
         <remarks>To be added.</remarks>
         <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="value" /> is an invalid Unicode scalar.</exception>
```

rpetrusha commented on xml/System.Text/Rune.xml:
```diff
@@ -1032,9 +1032,9 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
+        <param name="value">The 32-bit signed integer to convert.</param>
+        <summary>Defines an explicit conversion of a 32-bit unsigned integer to a <see cref="T:System.Text.Rune"/>.</summary>
+        <returns><paramref name="value" /> as a 32-bit Unicode character.</returns>
         <remarks>To be added.</remarks>
```
Exception for an invalid Unicode scalar value.

Jozkee replied:
```suggestion
         <remarks>To be added.</remarks>
         <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="value" /> is an invalid Unicode scalar.</exception>
```

Jozkee commented on xml/System.Text/Rune.xml:
```diff
@@ -977,9 +977,9 @@ This method is very similar to <xref:System.Text.Rune.DecodeFromUtf8(System.Read
+        <param name="ch">The 16-bit Unicode character to convert.</param>
+        <summary>Defines an explicit conversion of a 16-bit Unicode character to a <see cref="T:System.Text.Rune"/>.</summary>
+        <returns><paramref name="ch" /> as a 32-bit Unicode character.</returns>
         <remarks>To be added.</remarks>
```
```suggestion
         <remarks>To be added.</remarks>
         <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="ch" /> is a Unicode scalar that represents a surrogate code point.</exception>
```

Jozkee commented:
@rpetrusha I've addressed the feedback. Please tell me if something else is missing. Also plese verify the wording in the exceptions added to op_Explicit.

rpetrusha commented:
This looks good, @Jozkee. Thanks for making the additional changes. I'll merge now.
